### PR TITLE
⬆ Upgrade Starlette from 0.17.1 to 0.18.0

### PR DESCRIPTION
**Pull request by Kludex**
For reference, set of changes that broke FastAPI:
- https://github.com/encode/starlette/pull/1383 (mypy)
- https://github.com/encode/starlette/pull/1397
- https://github.com/encode/starlette/pull/1395

Related to #4477

### CODE REVIEW

File: fastapi/dependencies/utils.py [hunk 1]
```diff
@@ -462,13 +462,10 @@ async def solve_dependencies(
...
-        content=None,
-        status_code=None,  # type: ignore
-        headers=None,  # type: ignore # in Starlette
-        media_type=None,  # type: ignore # in Starlette
-        background=None,  # type: ignore # in Starlette
-    )
+    if response is None:
+        response = Response()
+        del response.headers["content-length"]
+        response.status_code = None
```
Comment (Kludex) [inline fastapi/dependencies/utils.py:468]: @tomchristie FastAPI uses **a** `Response` object that travels through the dependencies, but that `Response` is not the response that will be returned at the end. There's a "cleanup" after the dependencies are solved. You can see the "cleanup" logic below:

https://github.com/tiangolo/fastapi/blob/291180bf2d8c39e84860c2426b1d58b6c80f6fef/fastapi/routing.py#L216-L254

I've selected the whole snippet, from the point we get the `Response` object used in the dependencies (`solve_dependencies` call) until the last 5 lines (on which the values from `Response` are passed to the actual response object that will be used.

This comment is JFYK, I don't expect any action to be taken on Starlette. Also, considering that FastAPI tweaks Starlette internally for some features it has.

File: fastapi/routing.py [hunk 2]
```diff
@@ -126,7 +126,7 @@ async def serialize_response(
         if is_coroutine:
             value, errors_ = field.validate(response_content, {}, loc=("response",))
         else:
-            value, errors_ = await run_in_threadpool(
+            value, errors_ = await run_in_threadpool(  # type: ignore[misc]
```
Comment (Kludex) [inline fastapi/routing.py:129]: We added `ParamSpec` to `run_in_threadpool` on Starlette. The error here was:
```
fastapi/routing.py:129: error: "<nothing>" object is not iterable  [misc]
```

**Review by yezz123 — APPROVED** [review]
@Kludex 🌟

**Comment by L1ghtn1ng** [comment]
can this get merged soon and released please?

**Comment by ShadowLNC** [comment]
[Starlette 0.19.0](https://github.com/encode/starlette/releases/tag/0.19.0) has now been released, not sure if it's worth performing a further upgrade here, or leaving that for another time.

**Comment by Kludex** [comment]
As per history, Sebastián prefers to bump progressively. Also, those two releases have a lot of changes.

**Review by BilalAlpaslan — APPROVED** [review]

**Review by mdczaplicki — APPROVED** [review]

**Comment by Kludex** [comment]
> Hello ! I would like to know more regarding the status of this PR. Is there a reason not to merge it ?

Let's wait for @tiangolo :)

**Comment by tiangolo** [comment]
Awesome @Kludex, thank you! :bow: 

I added a filterwarning for pytest complaining about Cryptography so we can run tests while reaching the version that drops support for Python 3.6.

This will be available in the next release in a couple of hours, FastAPI `0.76.0` :label:

**Pull request merged**